File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
##.index()는 O(N)이라 오래걸림
##dic쓰면 O(1)이라 더 빠름
def solution(players, callings):
    player_dict = {player: idx for idx, player in enumerate(players)}
    
    for call in callings:
        idx = player_dict[call]
        
        players[idx], players[idx-1] = players[idx-1], players[idx]
        
        player_dict[players[idx]] = idx
        player_dict[players[idx-1]] = idx-1
    
    return players

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        Map<String,Integer> map=new HashMap<>();
        for(int i=0;i<players.length;i++){
            map.put(players[i],i);
        }
        
        for(String name: callings){
            int curIndex=map.get(name);
            if (curIndex==0){
                continue;
            }
            int prevIndex=curIndex-1;
            String name2=players[prevIndex];
            
            //swap
            players[prevIndex]=name;
            players[curIndex]=name2;
            
            map.put(name,prevIndex);
            map.put(name2,curIndex);
            
        }
        return players;
    }
}